membrane/client/flask/__init__.py

```diff
@@ -0,0 +1,228 @@
+
+
+def _create_custom_token(
+    redirect_url: str | None = None,
+    token_expiration: int | None = None,
+    custom_claims: dict | None = None,
+) -> str:
+    """Create a custom JWT token."""
+    _check_custom_claims(custom_claims or {})
+    payload = {
+        "app_id": _config.certificate.app_id,
+        "redirect_url": redirect_url or _redirect_url(),
+        "exp": _get_exp_date(token_expiration),
+    }
+    payload.update(custom_claims or _config.custom_claims)
+    headers = {"alg": ALGORITHM, "app_id": _config.certificate.app_id, "typ": "JWT"}
+    return _encode_jwt(payload, headers)
+
+
+def _encode_jwt(payload: dict, headers: dict) -> str:
+    """Encode JWT."""
+    try:
+        return jwt.encode(
+            payload,
+            _config.certificate.client_private_key,
+            algorithm=ALGORITHM,
+            headers=headers,
+        )
+    except jwt.InvalidTokenError as e:
+        raise JWTDecodingError(f"JWT decoding failed: {e}") from e
+
+
+def _verify_token(token: str) -> dict:
+    """Verify JWT token."""
+    try:
+        decoded_token = jwt.decode(
+            token, _config.certificate.server_public_key, algorithms=[ALGORITHM]
+        )
+        exp = decoded_token.get("exp")
+        if not exp:
+            raise JWTExpMissingError("Missing 'exp' field in the token.")
+        current_time = datetime.utcnow().timestamp()
+        if current_time > exp:
+            raise JWTExpiredError("The token has expired.")
+        return decoded_token
+    except jwt.InvalidTokenError as e:
+        raise JWTDecodingError(f"JWT decoding failed: {e}") from e
+
+
+def membrane_login_required(f):
+    """Decorator for login required routes."""
+
+    @wraps(f)
+    def decorated_view(*args, **kwargs):
+        if _config.require_login:
+            return _login_required(f)(*args, **kwargs)
+        return f(*args, **kwargs)
+
+    return decorated_view
+
+
+@_login_manager.user_loader
+def _load_user(user_id: str) -> User:
+    """Load user for Flask-Login."""
+    return User(user_id)
+
+
+@_login_manager.unauthorized_handler
+def _unauthorized():
+    """Handle unauthorized access."""
+    if (jwt_token := request.args.get("token")) is not None:
+        if (decoded_token := _verify_token(jwt_token)) is not None:
+            login_user(User(decoded_token["sub"]))
+            return redirect(url_for(request.endpoint, _external=True))
+    return _redirect_for_authentication()
+
+
+def _redirect_for_authentication():
+    """Redirect for authentication."""
+    jwt_token = _create_custom_token(request.url)
+    return redirect(f"{_config.certificate.auth_url}/authenticate?token={jwt_token}")
+
+
+@blueprint.route("/login")
+def login():
+    """Login route."""
+    if hasattr(membrane_current_user, "id"):
+        return redirect(_redirect_url())
+    return _redirect_for_authentication()
+
+
+@blueprint.route("/logout")
+@membrane_login_required
+def logout():
+    """Logout route."""
+    logout_user()
```

Comment:
doesn't this just clear the local cookie? since we're doing SSO, we probably want to logout on the membrane backend server (logout from one app logs out of all apps).

Reply:
Still looking into this.

Reply:
@k-allagbe as discussed, we can postpone the global logout issue to later (although bouncing systematically from landing page to membrane-backend could help ensure the token is valid)